messaging/tck/src/main/java/org/eclipse/microprofile/reactive/messaging/tck/arquillian/TckArquillianDeployListener.java

```diff
@@ -0,0 +1,141 @@
+import org.jboss.arquillian.test.spi.TestClass;
+import org.jboss.shrinkwrap.api.Archive;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.ServiceLoader;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class TckArquillianDeployListener {
+
+  private final TckContainer tckContainer;
+
+  private final Map<Class<?>, List<DeploymentDescription>> activeDeployments = new ConcurrentHashMap<>();
+
+  /**
+   * I don't know how deployments work in arquillian, I assume I need to explicitly deploy everything?
+   */
+  private final Archive<?> frameworkArchive = ShrinkWrap.create(JavaArchive.class)
+      .addClass(MockedReceiver.class)
+      .addClass(MockedSender.class)
+      .addClass(ReceiveTimeoutException.class)
+      .addClass(TestEnvironment.class)
+      .addClass(MockPayload.class)
+      .addClass(SimpleMessage.class)
+      .addClass(TckMessagingManager.class)
+      .addClass(ContainerController.class);
+
+
+  public TckArquillianDeployListener() {
+    Iterator<TckContainer> containers = ServiceLoader.load(TckContainer.class).iterator();
+
+    if (containers.hasNext()) {
+      this.tckContainer = containers.next();
+    }
+    else {
+      throw new RuntimeException("No " + TckContainer.class.getName() +
+          " found. To run this TCK, you must provide an implementation of " +
+          TckContainer.class + " via the JDK service loader mechanism.");
+    }
+  }
+
+  public void onBeforeDeploy(@Observes BeforeDeploy beforeDeploy, TestClass testClass) throws DeploymentException {
+    Topics topics = testClass.getAnnotation(Topics.class);
+    String[] topicNames;
+    if (topics != null) {
+      topicNames = topics.value();
+    }
+    else {
+      topicNames = new String[0];
+    }
+
+    DeployableContainer<?> container = beforeDeploy.getDeployableContainer();
+    DeploymentDescription description = beforeDeploy.getDeployment();
+
+    if (description.isArchiveDeployment()) {
+      description.getArchive().merge(frameworkArchive);
+    }
+
+    List<DeploymentDescription> deployments = tckContainer.createDeployments(topicNames);
+    List<DeploymentDescription> deployed = new ArrayList<>();
+    activeDeployments.put(testClass.getJavaClass(), deployed);
+    for (DeploymentDescription deployable: deployments) {
+      if (!tckContainer.mergeArchives() || deployable.isDescriptorDeployment() || description.isDescriptorDeployment()) {
+        deployed.add(deployable);
+        container.deploy(deployable.getDescriptor());
+      }
+      else {
+        description.getArchive().merge(deployable.getArchive());
+      }
+    }
+  }
+
+  public void onAfterUnDeploy(@Observes AfterUnDeploy afterUnDeploy, TestClass testClass) throws DeploymentException {
+    Topics topics = testClass.getAnnotation(Topics.class);
```

Comment:
I would add all this logic of getting the annotation and if/else into a private method to make the `onAfterUnDeploy` clear

Reply:
Will do.